src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -294,49 +325,62 @@ def _frobenius_charpoly_crystalline(self, var='X'):
-        n1, n0 = n // nstar, n % nstar
-        dm = self.coefficients(sparse=False)
-        S = PolynomialRing(L, name=str(A.gen()))
-        SM = MatrixSpace(S, r, r)
-        mu_coeffs = ((S.gen() - dm[0])**(n+1)).coefficients(sparse=False)
+        nquo, nrem = divmod(n, nstar)
+        drin_coeff = self.coefficients(sparse=False)
+        poly_K = PolynomialRing(K, name=str(A.gen()))
+        matrix_poly_K = MatrixSpace(poly_K, r, r)
+        mu_coeffs = ((poly_K.gen() - drin_coeff[0])**(n+1)).coefficients(
+                                                            sparse=False)
 
         def companion(order):
             # + [1] is required to satisfy formatting for companion matrix
-            M = SM(companion_matrix([(dm[i]/dm[r])**(q**order)\
-                    for i in range(r)] + [1], format = 'top'))
-            M[0, r-1] += S.gen() / dm[r]**(q**order)
+            M = matrix_poly_K(companion_matrix(
+                              [(drin_coeff[i]/drin_coeff[r])**(q**order)
+                               for i in range(r)] + [1], format='top'))
+            M[0, r-1] += poly_K.gen() / drin_coeff[r]**(q**order)
             return M
-        C0 = prod([companion(i) for i in range(n0, 0, -1)])
-        C = prod([companion(i) for i in range(nstar + n0, n0, -1)])
+
+        companion_initial = prod([companion(i)
+                                 for i in range(nrem, 0, -1)])
+        companion_step = prod([companion(i)
+                              for i in range(nstar + nrem, nrem, -1)])
         reduced_companions = []
-        for k in range(n1 - 1, 0, -1):
-            M = Matrix(S, r, r)
-            for i, row in enumerate(C):
+        for k in range(nquo - 1, 0, -1):
+            M = Matrix(poly_K, r, r)
+            for i, row in enumerate(companion_step):
                 for j, entry in enumerate(row):
-                    reduction = entry % S([c**(q**(-k*nstar % n))
-                                        for c in mu_coeffs])
-                    M[i, j] = S([c**(q**(k*nstar)) \
-                                for c in reduction.coefficients(sparse=False)])
+                    reduction = entry % poly_K([c**(q**(-k*nstar % n))
+                                                for c in mu_coeffs])
+                    M[i, j] = poly_K([c**(q**(k*nstar))
+                                     for c in reduction.coefficients(
+                                              sparse=False)])
             reduced_companions.append(M)
-        charpoly_coeffs_L = (prod(reduced_companions) * C * C0).charpoly(var)\
-                            .coefficients(sparse=False)
+        charpoly_coeffs_K = (prod(reduced_companions) * companion_step *
+                             companion_initial).charpoly(var).coefficients(
+                             sparse=False)
         # The above line obtains a char poly with coefficients in L[T]
         # This maps them into A
         return PolynomialRing(A, name=var)(
```

Comment:
For this block, I would suggest this:

Same number of lines, seems cleaner. 

```python
        def coeff_A(coeff):
            return A(list(map(lambda x: K(x).vector()[0], coeff)))

        coeffs_A = list(map(lambda coeff: coeff_A(coeff), charpoly_coeffs_K))
        return PolynomialRing(A, name=var)(coeffs_A)
```

Same number of (non blank) lines.